packages/gasket-intl/src/with-intl-provider.js

```diff
@@ -0,0 +1,114 @@
+import React, { useReducer } from 'react';
+import hoistNonReactStatics from 'hoist-non-react-statics';
+import merge from 'lodash.merge';
+import { IntlProvider } from 'react-intl';
+import { GasketIntlContext } from './hooks';
+import { clientData, isBrowser } from './config';
+import { LOADED, getActiveLocale } from './utils';
+
+/**
+ * Merges any initial state from render with that from page props
+ *
+ * @param {LocalesProps|{}} localesProps - Initial props from a Next.js page
+ * @returns {LocalesState} state
+ */
+export function init(localesProps) {
+  const { messages = {}, status = {} } = localesProps;
+  if (isBrowser) {
+    // merge any data set on window with what comes from SSR or static page props
+    const { messages: dataMessages = {}, status: dataStatus = {} } = clientData;
+    return merge({}, { messages: dataMessages, status: dataStatus }, { messages, status });
+  }
+
+  return { messages, status };
+}
+
+/**
+ *
+ * @param {LocalesState} state - Incoming state
+ * @param {{type: string, payload: {}}} action - State change action
+ * @returns {LocalesState} state result
+ */
+export function reducer(state, action) {
+  const { type } = action;
+  const { locale, messages, file } = action.payload;
+  if (type === LOADED) {
+    return {
+      ...state,
+      messages: {
+        ...state.messages,
+        [locale]: {
+          ...state.messages[locale],
+          ...messages
+        }
+      },
+      status: {
+        ...state.status,
+        [file]: LOADED
+      }
```

Comment:
should the messages be separated by file? if there's 2 files with the same message key, are they going to stomp each other's message?

Reply:
should this all be
```
return {
  ...state,
  [file]: {
     ...state[file],
     messages: {
        ...etc
     },
     status: LOADED/type/etc
   }
}
```

Reply:
No, we need the combined messages for a locale, which is passed down to the react-intl provider. Yes, there is a risk of stomping over messages with the same key, but that seems like a rare edge case and more of a content problem for the app dev though. We could add a check and log a warning here, but it feels like static analysis of the JSON files would be most appropriate for teams. For example [eslint-plugin-i18n-json](https://github.com/godaddy/eslint-plugin-i18n-json) can do this.

Reply:
ah okay. 👍